File: ObserverV2C#/Observer/src/ClockTimer.cs
using Observer.include;
using System;
using System.Collections.Generic;
using System.Text;

namespace Observer.src
{
    public class ClockTimer: IClockTimer {

        public ClockTimer(ISubscriber subscriber) {
            notifier = new Notifier(subscriber);
            hours = 0;
            minutes = 0;
            seconds = 0;
            this.subscriber = subscriber;
        }

        public int GetHour()
        {
            return this.hours;
        }

        public int GetMinute()
        {
            return this.minutes;
        }

        public int GetSecond()
        {
            return this.seconds;
        }

        public ISubscriber GetSubscriber()
        {
            return this.subscriber;
        }

        public void Tick()
        {
            seconds++;
            if (seconds == 60)
            {
                seconds = 0;
                minutes++;
                if (minutes == 60)
                {
                    seconds = 0;
                    minutes = 0;
                    hours++;
                    if (hours == 24)
                    {
                        seconds = 0;
                        minutes = 0;
                        hours = 0;
                    }
                }
            }
            notifier.Notify();
        }




        private int hours;
        private int minutes;
        private int seconds;
        private ISubscriber subscriber;
        private INotifier notifier;
    }
}

File: ObserverV3/src/ClockTimer.cpp
#include "../src/Notifier.cpp"
#include "../include/IClockTimer.h"
#include "../include/SmartPointer.h"

class ClockTimer: public IClockTimer
{
    public:
    ClockTimer(ISubscriber* subscriber): subs(subscriber),notifier(new Notifier(subscriber)) {
            m_hours = 0;
            m_minutes = 0;
            m_seconds = 0;
            std::cout << "ClockTimer::ClockTimer" << std::endl;
        }

        virtual ~ClockTimer(){
            std::cout << "ClockTimer::~ClockTimer" << std::endl;
        }

        int getHour( void ) {
            return this->m_hours;
        }

        int getMinute( void ) {
            return this->m_minutes;
        }

        int getSecond( void ) {
            return this->m_seconds;
        }

        ISubscriber* getSubscriber() {
            return  subs;
        }

        void tick () {
            m_seconds++;
            if(m_seconds == 60) {
                m_seconds = 0;
                m_minutes++;
                if(m_minutes == 60) {
                    m_seconds = 0;
                    m_minutes = 0;
                    m_hours++;
                    if(m_hours == 24) {
                        m_seconds = 0;
                        m_minutes = 0;
                        m_hours = 0;
                    }
                }
            }
            notifier->notify();
        }

    private:
        int m_hours;
        int m_minutes;
        int m_seconds;
        ISubscriber* subs;
        Smart_ptr<INotifier> notifier;
};


